src/mpid/ch4/netmod/ofi/ofi_init.c

```diff
@@ -1003,13 +1003,20 @@ int MPIDI_OFI_mpi_finalize_hook(void)
 {
     int thr_err = 0, mpi_errno = MPI_SUCCESS;
     int i = 0;
-    int barrier[2] = { 0 };
-    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
     MPIR_Comm *comm;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_OFI_FINALIZE);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_OFI_FINALIZE);
 
+    /* There is no way to ensure remote-completion of fi_inject. If we proceed
+     * tearing down here, the remote recv may be left hanging. Use barrier to
+     * synchronize, but force non-immediate send so MPIDI_OFI_PROGRESS may
+     * ensure send completion) */
+    MPIR_Errflag_t errflag = MPIR_ERR_NONE;
+    MPIDI_OFI_global.max_buffered_send = 0;
+    mpi_errno = MPIR_Barrier_fallback(MPIR_Process.comm_world, &errflag);
```

Comment:
To me, this less explicit than the previous code + comment. The previous could still be improved if we state that each process is contributing an `MPI_INT` to `MPIR_Allreduce` to go beyond the `fi_inject` threshold. The "fallback" barrier is masking that complexity. Maybe there is another name that would help?

Reply:
I see. 

There are 3 points I was addressing with the change. 1. We needed a "barrier" here to make sure none of the process proceed to tear down before the send is remote complete. The previous "allreduce" is a bit obscure in that purpose. 2. I would like to call a "fallback" version of collective rather than an "auto" version due to the unpredictable and evolving nature of the latter. Even the "fallback" version is an exact duplicate of one of the algorithm, it still provides the isolation of the dependency. 3. We don't want to introduce additional pending send-completions. The `max_buffered_send=0` didn't work for message size 0, is really a side effect that we use unsigned type for `max_buffered_send`. If we used a signed version such as `MPI_Aint`, we can simply set it to `-1`. Because switching to signed types are more extensive, I chose to modify the "Barrier" code instead. It is really a side-effect that is not related to either the "barrier" part nor the "fallback" part.

The comment I added is to explain why we need a synchronization code here at all, as I did not find the reason obvious.

I see your point. I guess I could add another line of comment explaining that `MPIR_Barrier_fallback` uses finite-sized message therefore the setting `max_buffered_send=0` works here.

Reply:
I should say I'm not opposed to changing this in principal. The existing method is a hack, but it is effective. My issue with this version is the clarity. "Fallback" to me is something used when another method will not work. If we are clear about _why_ a regular `MPIR_Barrier` may fail in this case and how the fallback ensures success, then I'm OK with the change.

In reality, I still think properly releasing ofi resources should ensure `fi_inject` messages will not be silently dropped at the source. If we are not doing something correctly in that regard, we should fix it. If we are doing things correctly, perhaps we can contribute a test to libfabric to smoke out issues with various providers.

Reply:
I see that we are actually hanging up on the choice of word -- "fallback".  I have been associating it with dependency as the "fallback" should work when "regular" may not work (due to dependency). But I see your view point. 

In this case, `MPIR_Barrier` can be made work -- my preferred fix is to apply `MPI_Aint` PR and change the threshold here to `-1`. But even then I still would like to call a specific "init-safe" version so `MPIR_Barrier` may be free to evolve while leaving the init/finalize stable. The case in point is, `MPI_Allreduce` was working until some developer evolved the selection or algorithm that has more dependency. Such situation is tricky because ideally the people who maintain ofi init/finalize should not be burdened to understand how `MPI_Allreduce` work and its dependency and vice versa. A fallback version relieves that burden. 

Let me know if you agree that we should have a separate "stable" version in init/finalize here instead of calling a run-time "auto" version collectives. 

If we agree above, then we can move on to the name itself. If you don't like "fallback", any suggestions?

> In reality, I still think properly releasing ofi resources should ensure fi_inject messages will not be silently dropped at the source. If we are not doing something correctly in that regard, we should fix it. If we are doing things correctly, perhaps we can contribute a test to libfabric to smoke out issues with various providers.

I agree with you here. One of the `fi_close` should be blocking calls wait for all message out-of-nic or exit with a timeout. But on the other hand, have a synchronization call at the beginning of `MPI_Finalize` ensures a clean state is not a bad practice -- just like a defensive code with double ensurance. In addition, we'll call `MPI_Allgather` in init, which result in a similar situation. I would like to call a "fallback" version there as well.

Reply:
> I see that we are actually hanging up on the choice of word -- "fallback". I have been associating it with dependency as the "fallback" should work when "regular" may not work (due to dependency). But I see your view point.
> 
> In this case, "fallback" does not mean `MPIR_Barrier` won't work. `MPIR_Barrier` can be made work -- my preferred fix is to apply `MPI_Aint` PR and change the threshold here to `-1`.

I think this is a good idea.

> But even then I still would like to call a specific "init-safe" version so `MPIR_Barrier` may be free to evolve while leaving the init/finalize stable. The case in point is, `MPI_Allreduce` was working until some developer evolved the selection or algorithm that introduced more dependency. Such situation is tricky because ideally the people who maintain ofi init/finalize should not be burdened to understand how `MPI_Allreduce` work and its dependency and vice versa. A fallback version relieves that burden.
> 
> Let me know if you agree that we should have a separate "stable" version in init/finalize here instead of calling a run-time "auto" version collectives.

Sounds reasonable to me.

> 
> If we agree above, then we can move on to the name itself. If you don't like "fallback", any suggestions?

Hmm, nothing jumps out at me as obvious. Maybe "simple" or "basic"?

> 
> > In reality, I still think properly releasing ofi resources should ensure fi_inject messages will not be silently dropped at the source. If we are not doing something correctly in that regard, we should fix it. If we are doing things correctly, perhaps we can contribute a test to libfabric to smoke out issues with various providers.
> 
> I agree with you here. One of the `fi_close` should be blocking calls wait for all message out-of-nic or exit with a timeout. But on the other hand, have a synchronization call at the beginning of `MPI_Finalize` to ensure a synchronized state is not a bad practice -- it simplifies the code assumptions and it adheres to the standard. Also, we need call collective in init anyway (`MPI_Allgather`), which result in a similar situation. I would like to call a "fallback" version there as well.

Sure. I know we had this discussion in the ofi working group before and the barrier has remained in MPICH, so maybe it's the best we can do for now.